### ISSUE TITLE
Pixi: fail to use renderer

### ISSUE THREAD

**Issue by whitetigle** [fable-compiler/fable-graphics#1]
Hi!
I am trying to setup a very simple Pixi project. However I fail right at the start with a compilation error.
Since I am new to both Fable & F# there may be things I do wrong.  Or maybe I did not understand the **Erase attribute** sample in the doc since **Globals.autodetectRenderer** returns a **U2<WebGLRenderer, CanvasRenderer>**

Anyway here is my code : 

```
open System
open Fable.Core
open Fable.Core.JsInterop
open Fable.Import.PIXI
open Fable.Import.Browser

let currentRenderer =
  let renderer = Globals.autoDetectRenderer( 800., 600.  )
  match renderer with
    | U2.Case1 webgl -> webgl :> SystemRenderer
    | U2.Case2 canvas -> canvas :> SystemRenderer

document.body.appendChild( currentRenderer.view )
```

When I compile it says : 

> fable-compiler 0.5.8: Start compilation...
> ERROR: Unsupported type test: GenericParam "b" (L27,8-L27,16) 

which corresponds to my renderer value

> match **renderer** with

I use fable-core 0.5.5. 
I have tried other ways but I can't get past this error. 
Thanks for your kind help!

**Comment by alfonsogarciacaro** [comment]
Hi there! The Pixi bindings aren't very tested so there may be some rough edges, sorry for that!

The erased union types are indeed erased in runtime that's why you cannot do type testing against them. I'm not sure if it'll solve the problem but probably the best thing is to cast the returned valued directly with `unbox`.

``` fsharp
let currentRenderer =
  Globals.autoDetectRenderer( 800., 600.  )
  |> unbox<SystemRenderer>
```

**Comment by whitetigle** [comment]
Ok that works and it's simpler. Thanks!

Regarding the Erased unions, I was following the example from https://fable-compiler.github.io/docs/interacting.html page:

```
open Fable.Core

type Test() =
    member x.Value = 'Test'


let myMethod (arg: U3<string, int, Test>) =
    match arg with
    | U3.Case1 s -> s
    | U3.Case2 i -> string i
    | U3.Case3 t -> t.Value
```

Now you gave me the answer, I just tried to compile this example and it gives the same error I got. I should have started with this and ask you to understand what was happening.

So to summarize we don't have to do the whole matching but we can just unbox to the desired type. Is that so? 

Now, I have done quite a few projects using Pixi in another language (Haxe). I am actually trying to port one using Fable and F# so If I encounter other problems may I ping you from time to time? 
If so is is better to use gitter for that or submit an issue? 

Anyway thanks for your help!

**Comment by whitetigle** [comment]
I put my notes here while I am working (please tell me if you want to start another discussion elsewhere).

**1. optional values VS default values**

Let's take an example. Setting optional parameters for the renderer.

JS
`var renderer = PIXI.autoDetectRenderer(800, 600, { antialias: true });`

F#

```
type Options() =
  interface RendererOptions with
    member val antialias = Some(true) with get, set
    member val view = None with get, set
    member val transparent = None with get, set
    member val autoResize = None with get, set
    member val resolution = Some(1.) with get, set
    member val clearBeforeRendering = None with get, set
    member val preserveDrawingBuffer = None with get, set
    member val forceFXAA = None with get, set
    member val roundPixels = None with get, set
    member val backgroundColor = Some(float 0x000000) with get, set

let currentRenderer =
  Globals.autoDetectRenderer( 800., 600., Options() )
  |> unbox<SystemRenderer>
```

If I understand correctly interfaces in F# we must explicitely define all members. So setting a None value when the JS code is waiting for, let's say, a float value is certainly not what we want to do..

*_So here is my question: is there a way to define default values instead or make options becomes REALLY optional (create the object in JS without the field) *_

For instance, the resolution parameter of my Options object is 1.0 by default. But here since we use an interface, we must explicitely set Some(1.0) since None is not what we want.

In JS we just omit the parameter. In Haxe, we omit the field and it will create an anonymous object without the parameter. Can we achieve the same in F#? Am i right regarding the use of interfaces in F#?

Thanks

**Comment by whitetigle** [comment]
Fable Samples using Pixi:
Pixi graphics API use sample: https://gist.github.com/whitetigle/f1bd57859ce4a3f9fd9155f38047cfca

Instead of porting my app, The best move I can do now is to port the official samples so we check everything is working. 
So for now I will put them in a public repository so we can double check the code is clean and is following fable guidemines and then if you wish we'll be able to add them somewhere in the Fable repository in a samples section.

**[EDIT]**: just found the samples dir https://github.com/fable-compiler/Fable/tree/master/samples/browser/pixi

However since there are a lot of samples, I wonder how to define a single build file to generate the .js files from all the .fsx files located in a folder. I have not yet found a way to do this easily using fableconfig.json (any hint in the doc?) Do I need to use a .fsproj file? (excuse these newbie questions)

**[EDIT 2]** I will go for a simple system script for now :)

**Comment by whitetigle** [comment]
So far so good. Went up the 5 first samples without any problem. (although it may not be the best f# code one can read since it's quite new for me)

Anyway I have prepared a simple html page that lists the samples and loads each of them dynamically showing both the f# source code and the sample. It's not live yet (in a private repository). I will go through all the samples and then we'll see what to make of all this codebase. 

Screenshot: https://postimg.org/image/coogcm4c5/

**Comment by alfonsogarciacaro** [comment]
Sorry for the late reply. This is an awesome job! :clap: It'd be fantastic to add the samples to Fable's website 😄  I'll try to answer your questions now:

> Regarding the Erased unions, I was following the example from https://fable-compiler.github.io/docs/interacting.html page:

Thanks for pointing this out! This was actually a bug in Fable, it should be possible to type test erased union types (e.g. `match x with U2<string,int>.Case1 s -> ...` will translate to `typeof x === "string" ? ...`). It will be fixed in the next release :+1: However, in your specific case I think unboxing is fine. Of course `unbox` is less safe than a type test, but there both branches were casting to `SystemRenderer` so it should be fine.

> If I encounter other problems may I ping you from time to time? If so is is better to use gitter for that or submit an issue?

Of course you can ping me :) Gitter is good for short questions and there're also other people listening there so you may get an answer faster there. But GH issues are also good to record answers so other people can easily find the info. Whatever is more convenient for you will work.

> Is there a way to define default values instead or make options becomes REALLY optional (create the object in JS without the field)

Yeah, this is an issue as unfortunately there's no way in F# to have optional members 😕 But as most of the times in JS this is done for option objects (as in this case) we have an alternate way: [key value lists](https://fable-compiler.github.io/docs/interacting.html#KeyValueList-attribute). Unfortunately, `ts2fable` won't create this type of definitions but it's easy to do it yourself:

``` fsharp
[<KeyValueList>]
type RendererOptions =
    | View of HTMLCanvasElement
    | Transparent of bool
    | Antialias of bool
    | AutoResize of bool
    | Resolution of float
    | ClearBeforeRendering of bool
    | PreserveDrawingBuffer of bool
    | ForceFXAA of bool
    | RoundPixels of bool
    | BackgroundColor of float
```

Now you can only specify the options you need:

``` fsharp
let currentRenderer =
    let options = [
        Antialias true
        Resolution 1.
        BackgroundColor (float 0x000000)
    ]
    Globals.autoDetectRenderer( 800., 600., options )
    |> unbox<SystemRenderer> 
```

> However since there are a lot of samples, I wonder how to define a single build file to generate the .js files from all the .fsx files located in a folder. I have not yet found a way to do this easily using fableconfig.json (any hint in the doc?) Do I need to use a .fsproj file?

It depends if you consider all samples to be part of the same project (in this case we may think of a project as one HTML page). `fableconfig.json` doesn't support multiple projects but if you want to compile all samples at once in the same project, you can use `.fsproj` or just another script that loads the others with the `#load` directive ([example](https://github.com/fable-compiler/fable-electron/blob/master/samples/helloworld/src/files.fsx)). For the websites samples we usually use a script file with a tutorial and a small configuration to automatically set up the web page. See this issue for an explanation of the process: https://github.com/fable-compiler/Fable/issues/162

I hope this helps! Please tell me if you have more questions and thanks a lot for your contribution!

**Comment by whitetigle** [comment]
Hi! 
Thanks for your feedback. So here I am with some design problems: I am trying to find a **safe way to pass a parameter that uses default values from JS**.

context: http://pixijs.github.io/examples/index.html?s=demos&f=movieclip-demo.js&title=MovieClip#/basics/render-texture-v3.js

So in the JS sample there's the following piece of code 

``` javascript
var rt = new PIXI.RenderTexture(renderer, 300, 200, PIXI.SCALE_MODES.LINEAR, 0.1);
```

So as you can see we've got this **PIXI.SCALE_MODES.LINEAR** value which comes from pixi.js

``` javascript
    SCALE_MODES: {
        DEFAULT:    0,
        LINEAR:     0,
        NEAREST:    1
    }
```

So I still have this problem regarding default values and how to make sure we don't end up putting some wrong values in critical parts of the applications especially when these values already exist in JS

So in our Pixi bindings our SCALE_MODES gets converted to 

``` f#
    type ScaleModes =
        abstract member DEFAULT: float
        abstract member LINEAR: float
        abstract member NEAREST: float
```

F# 1st way : way too complicated in my opinion and clearly not safe

``` f#
type ScaleMode() =
  interface ScaleModes with
    member this.LINEAR = 0.
    member this.DEFAULT = 0.
    member this.NEAREST = 1.

let mode = new ScaleMode()
let linear = mode :> ScaleModes

let rt = new RenderTexture(U2.Case2 renderer, 300., 200., linear.LINEAR, 0.1);
```

F# 2nd way : same result and really not safe

``` f#
let LINEAR = 0. // this is the value as found in pixi.js code
let rt = new RenderTexture(U2.Case2 renderer, 300., 200., LINEAR, 0.1);
```

F# 3rd way : safe as long as we don't emit wrong code. Too slow when it comes to testing: we must compile and then check at runtime

``` f#
[<Emit("PIXI.SCALE_MODES.LINEAR")>]
let LINEAR : float = failwith "JS only"
let rt = new RenderTexture(U2.Case2 renderer, 300., 200., LINEAR, 0.1);
```

I did not use Key Value lists since in this case it would not help better

Of course, I may miss other ways to do this since I miss experience in F#. 
That's why I would like to know what approach you would choose? Any other safer way?
Thanks!

**Comment by MangelMaxime** [comment]
We could used the third solution like this:

``` fsharp
module PIXI =

  type SCALE_MODES =
    [<Emit("PIXI.SCALE_MODES.LINEAR")>]
    static member LINEAR : float = failwith "JS only"
```

Here is how we can access it:

``` fsharp
let access = PIXI.SCALE_MODES.LINEAR
```

Which is transparent with Pixi way from JavaScript point of view.

@whitetigle I am also interested in Pixi.JS binding. If you want we could discuss on Gitter private channel to make our effort coordinate.

**Comment by whitetigle** [comment]
Hi @MangelMaxime 
I am currently evaluating using Fable/F# on my upcoming projects. Since I do a lot of things in Pixi, I wanted to try if I could port the samples easily. A way for me to get better at F# and see if the functional approach would allow for more robust apps as well as lighter codebase.

So far, my F# experiments have proven much more concise and robust that my previous attempt at functional code style using scala. Also the code is much more easier to read.

So since I have some time to dig further these days, my next goal is to port one of my former Pixi project made in Haxe and evaluate the Pros & Cons of this approach.

So for starters, with the help of @alfonsogarciacaro, I have started to port the official samples from Pixi.js; I am nearly done with the 12 samples from the **basics section** (http://pixijs.github.io/examples/index.html?s=demos&f=movieclip-demo.js&title=MovieClip#/basics/basic.js)

Since the v4 just came out, of course I will be very interested to work with it and if we can make the codebase more robust with solutions like you mentioned then it will be a win for everybody.

I must repeat but I am a F# newbie so these days I often stumble on basic stuff mostly things I haven't  yet played with. (like **fun vs Func** this morning) But I learn fast.

So like I said to @alfonsogarciacaro I think can help here by porting the samples and post issues / thoughts on the bindings to eventually have a clean, readable and robust codebase.

Now regarding the V4, since it's very new, I think that we should keep a v3 and v4. The v3 has been there for quite a long time so it would allow non F#/Fable users - if such audience exists -  to speed up their porting and it would also allow for a fallback.

I say this because this morning when working on the **TextStyle** object I saw that the way to declare a font has changed leading to breaking changes. So we need to be careful with this to avoid at all costs errors coming from wrong bindings. I wrote/used several bindings in Haxe and Scala so I know how much time one can lose to eventually find that the bindings where plain wrong.

Anyway, I will be glad to help as much as I can and also ask for help when needed.
(we can continue on gitter later if you want)

++

**Comment by alfonsogarciacaro** [comment]
In general I think it's not a good idea to access modules directly with `EmitAttribute` because sometimes Fable may change the name of the module internally to avoid conflicts. In any case, you don't need to implement the interface here, you should be able to do something like `PIXI.Globals. SCALE_MODES.LINEAR` (see [this](https://github.com/fable-compiler/fable-graphics/blob/master/pixi/Fable.Import.Pixi.fs#L1278)) Can you check if that works for you?

Please note that, because of the way TypeScript definitions are usually written, "static" values in the bindings won't be represented as static members in F# but just as a normal interface and then you can access an instance of this interface in the `Globals` wrapper. You can find more info in [ts2fable README](https://www.npmjs.com/package/ts2fable) on how the bindings work and why this `Globals` wrapper is needed.

**Comment by whitetigle** [comment]
PIXI.Globals. SCALE_MODES.LINEAR works just fine. Thanks! 
I'll check the ts2fable README.

**Comment by MangelMaxime** [comment]
@whitetigle Ok.

And thanks for showing the breaking change. I will play a little with v3 so in order to used a more "mature" lib. And will migrate later when confident with Pixi. :)

**Comment by whitetigle** [comment]
I just set my repository to public.
So if you wish you can use/check my samples to check: https://bitbucket.org/fnicaise/fs_press
- src folder contains the .fsx scripts
- out folder contains the .js outputs
- `make` to build the .js (I needed a simple build tool)

you can serve the files from the root and see the samples + their relative .fsx
It's not exactly clean, but it works for starters.

Note: Bitbucket is my by default source code server to play. So I started there. But I can move to github if needed early.

**Comment by MangelMaxime** [comment]
That really cool thanks. :)

I just needed to change the name in the package.json file because it's wasn't installing the dependencies as it's not compliant with the convention I think. `"name": "pixi-fsharp-samples",`

**Comment by whitetigle** [comment]
Ah yes, sorry, edited that. Thanks!

I have 2 new questions regarding the current sample:
http://pixijs.github.io/examples/index.html?s=demos&f=movieclip-demo.js&title=MovieClip#/basics/custom-filter-v3.js

**1 - createObject syntax**

I needed to create an uniform like in the sample and since the function was waiting for an obj I ended up writing this

``` f#
  let customUniform = createObj [
                        "customUniform" ==>
                            createObj [
                              "type" ==> "1f"
                              "value" ==> "0"]
                        ]
```

While it works, is this the relevant syntax or is there another way to write objs?

**2  - casting an obj to a string**
I am loading the fragment shader file and so I get a resources object with the data inside. I end up with:

``` f#
let fragmentSrc = resources?shader?data
```

I then cast the data to a string like this:

``` f#
let fragmentSrcString : string = sprintf "%O" fragmentSrc
```

Is this the correct way to do this in F#? 

[EDIT]
**3 - ResizeArray shortcut**
Is there a shorter way to create/use a ResizeArray ? 

``` f#
let filters = new ResizeArray<AbstractFilter>()
filters.Add(filter)
bg.filters <- filters
```

``` javascript
bg.filters = [filter];
```

**4 - change mutable value from an obj**
I did not find how to port this piece of code:

``` javascript
filter.uniforms.customUniform.value += 0.04;
```

in F# filter.uniforms is a property then how can I change the value of its inner object (if that's an object?) I was thinking about reading then writing the value like this 

``` f#
let curValue = filter?uniforms?customUniform?value
filter?uniforms?customUniform?value <- curValue + 0.04
```

But curValue being an obj I can not add it to a float (I know it is trivial but I can't find it in the docs) 

Thanks!

**Comment by alfonsogarciacaro** [comment]
Fantastic work, @whitetigle! I had a look at the samples and they look really good 👏 As you commented, the "least elegant" part is option declaration. So I just pushed I new version (0.0.4) of the Pixi bindings that replaces the interfaces with the `KeyValueList` defined above. Could you give it a try by following the instructions in my previous comment?

> 1 - createObject syntax

Yes, that's the correct syntax. It may be a bit verbose but I wanted to avoid conflicts. You can still define your own operators to make it terser as in [this example](https://github.com/fable-compiler/fable-electron/blob/master/samples/material-ui/src/renderer.fsx#L24-L30) (inlining the operators avoids unnecessary function calls in the generated JS code).

> 2 - casting an obj to a string

That syntax is correct, but if you just want to convert the object to string with no additional formatting you can simply use the `string` operator:

``` fsharp
let fragmentSrcString = string resources?shader?data
```

If you just want to **cast** the object to string. That is, you're sure the object is a string and you want to tell that the compiler (of course, this is less safe) `unbox` will also do:

``` fsharp
let fragmentSrcString = unbox<string> resources?shader?data
// let fragmentSrcString: string = unbox resources?shader?data // This is equivalent syntax
```

> 3 - ResizeArray shortcut

Yeah, ResizeArray can be a bit verbose. Actually, both F# `ResizeArray` and fixed arrays compile to the same JS array but the `ts2fable` parser must pick one so it picks `ResizeArray` by default. However you can construct `ResizeArray` directly from a literal array:

``` fsharp
bg.filters <- ResizeArray [|filter|]
```

(It's a bit more performant if you use a literal array but it works with any `seq` collection.)

**Comment by alfonsogarciacaro** [comment]
> 4 - change mutable value from an obj

F# doesn't include the `+=` to discourage this kind of mutations but they're indeed needed when dealing with some JS libs. Every time you're dealing with dynamic programming and the compiler complains `unbox` is your friend :)

``` fsharp
let curValue = unbox<float> filter?uniforms?customUniform?value
filter?uniforms?customUniform?value <- curValue + 0.04
```

**Comment by whitetigle** [comment]
I just retrieved your latest version but compilation fails: 

```
>Unexpected keyword 'and' in definition (L395,4-L395,7) (Fable.Import.Pixi.fs)
> Incomplete structured construct at or before this point in implementation file. Expected incomplete structured construct at or before this point or other token. (L427,8-L427,14) (Fable.Import.Pixi.fs)

```

I tried to change this but ionide crashes Atom each time I load such a _**big**_ file. Will try in VS.

**Comment by whitetigle** [comment]
Error is right there (line395)

``` f#
[<KeyValueList>]
    and RendererOptions =
        | View of HTMLCanvasElement
        | Transparent of bool
        | Antialias of bool
        | AutoResize of bool
        | Resolution of float
        | ClearBeforeRendering of bool
        | PreserveDrawingBuffer of bool
        | ForceFXAA of bool
        | RoundPixels of bool
        | BackgroundColor of float
```

[EDIT]
tried the following syntax and it compiles

```
and [<KeyValueList>]
        RendererOptions =
        | View of HTMLCanvasElement
        | Transparent of bool
        | Antialias of bool
        | AutoResize of bool
        | Resolution of float
        | ClearBeforeRendering of bool
        | PreserveDrawingBuffer of bool
        | ForceFXAA of bool
        | RoundPixels of bool
        | BackgroundColor of float

```

Going to check the samples now.

**Comment by whitetigle** [comment]
OK with patch applied it works! Bravo @alfonsogarciacaro  @! :+1: 
I am going to update all the samples now.

[EDIT] tried unbox and ResizeArray and it works. Thanks!

**Comment by alfonsogarciacaro** [comment]
Damn! I should have checked better before publishing, sorry for that! 😅 I fixed the error following your indications and republished the bindings (0.0.5).

Is Atom crashing with a file of that size? I'd recommend Visual Studio Code, it has more or less the same features and Ionide is available for it too :+1:

**Comment by whitetigle** [comment]
Well I do have VS Code somewhere so I will give it a try. Thanks!

**Comment by whitetigle** [comment]
Ok, here I am for some new questions (sorry!).

In javascript I have 

``` javascript
button.on('mousedown', onButtonDown)
function onButtonDown()
{
    this.alpha = 1;
}
```

where **this** is the current button.

Here is my current attempt in F# which fails

``` f#
let onButtonDown context  =
  let button = unbox<Sprite> context?button
  button.alpha <- 1.

let context = createObj[ "button" ==> button ]
button?on("mousedown", fun e -> onButtonDown(context)) |> ignore

```

it fails to capture the local context and thus is applied to the last button I set (I create buttons in a loop). Passing the button object leads to the same result.

So how can I do this in F#? 
Thanks!

**Comment by alfonsogarciacaro** [comment]
Argh! I hate when JS libs do funny things with `this`, that was the reason I have to give up the `Vue` bindings for Fable. Can you point me to the full JS source?

**Comment by whitetigle** [comment]
http://pixijs.github.io/examples/index.html?s=demos&f=movieclip-demo.js&title=MovieClip#/demos/interactivity.js

**Comment by whitetigle** [comment]
In Haxe we do keep the local context. For instance:

``` haxe
button.click = function( data ) {
  doStuff(button);
}   
```

But the first years (back in 2008 I think) we had to create a local context object and pass this object.

I tell you this because the Haxe compiler is made in OCAML so maybe there are things to get from here? (Not a compiler guy myself. It's just a thought)

**Comment by alfonsogarciacaro** [comment]
If you can attach the event handlers directly to the object, deriving may work. Could you please try this?

``` fsharp
type MyButton(texture) =
    inherit PIXI.Sprite(texture)
    member this.mousedown() =
        this?isdown <- true
        this?alpha <- 1.  
```

**Comment by whitetigle** [comment]
Well... Bravo! :+1: 
![ohyeah](https://cloud.githubusercontent.com/assets/883207/17978280/7ba02350-6af5-11e6-92fe-6f6642f52cbf.png)

**Comment by whitetigle** [comment]
Hi! Last question for today!

I have started to port this sample: http://pixijs.github.io/examples/index.html?s=demos&f=movieclip-demo.js&title=MovieClip#/demos/text-demo.js

Here we mix Text and BitmapTextObjects. So I started fiddling with our Pixi Bindings to use KeyValueList instead of long interfaces which led to:

``` f#
  and [<KeyValueList>] 
        TextStyle =
        | Font of string
        | Fill of U2<string,float>
        | Align of string
        | Stroke of U2<string,float>
        | StrokeThickness of float
        | WordWrap of bool
        | WordWrapWidth of float
        | LineHeight of float 
        | DropShadow of bool 
        | DropShadowColor of U2<string, float> 
        | DropShadowAngle of float 
        | DropShadowDistance of float 
        | Padding of float 
        | TextBaseline of string 
        | LineJoin of string 
        | MiterLimit of float 
```

I did the same for BitmapTextStyle and ended up with :

``` f#
        [<KeyValueList>]
        type BitmapTextStyle =
            | Font of U2<string, obj>
            | Align of string 
            | Tint of float 
```

I then modified parameter accesses whenever it was needed from TextStyle to TextStyle list and the same for BitmapTextstyle.

However F# does not seem to like having KeyFontValue types with Keys having the same name thus leading to errors when I try to do this:

``` f#
// BitmapTextStyle
  let fontStyle = [
    Font (U2.Case1 "35px Desyrel")
    Align "right"
  ]
  let mutable bitmapFontText = new extras.BitmapText("pixi bitmap fonts in f# are\n now supported", fontStyle)

// TextStyle
let style = [
  Font "35px Arial"
  Fill (U2.case1 "white")
  Align "left"
]
let mutable textSample : Fable.Import.PIXI.Text = Fable.Import.PIXI.Text("Rich text with a lot of options and across multiple lines", style)

```

In fact my problem is that the compiler does not know how to make the difference between both styles.  So what can I do?
Thanks!

(updated bindings here: https://gist.github.com/whitetigle/b03e4689823d2950b711a81d4793df04)
[EDIT] sample source code there: https://bitbucket.org/fnicaise/fs_press/raw/1cff44eb7ab90c2d90f20ac53810096c3b2d7e3f/src/demos/text-demo.fsx

**Comment by alfonsogarciacaro** [comment]
This is usually a problem when you have many union cases open in the same namespace and, most inevitably, some names conflict. In those cases, the compiler takes the last name defined. There are a couple of solutions for that:
- Qualify the case name with the name of the type:

``` fsharp
let style = [
  TextStyle.Font "35px Arial"
  TextStyle.Fill (U2.Case1 "white")
  TextStyle.Align "left"
]
```

That's a bit more verbose but you can always use type aliases:

``` fsharp
type TS = TextStyle

let style = [
  TS.Font "35px Arial"
  TS.Fill (U2.Case1 "white")
  TS.Align "left"
]
```

Also, if needed it's possible to force the user to always qualify the case name by adding `RequireQualifiedAccessAttribute` to the type:

``` fsharp
[<KeyValueList; RequireQualifiedAccess>] 
type TextStyle =
    | Font of string
    | Fill of U2<string,float>
    | Align of string
    ...
```
- The second solution is a bit more convoluted but it may be useful in some scenarios. You can change one of the colliding names to avoid the conflict but use `CompiledNameAttribute` to be sure the proper name will be used in the compiled JS code:

``` fsharp
[<KeyValueList>]
type BitmapTextStyle =
    | [<CompiledName("font")>] BFont of U2<string, obj>
    | [<CompiledName("align")>] BAlign of string 
    ...
```

> Note that in this case the compiled name should start with lower case (this is done automatically for _normal_ `KeyValueList` union cases).

**Comment by whitetigle** [comment]
Hi @alfonsogarciacaro ! Thanks for your help. SoI just used type aliases and it works.

I wonder about **RequireQualifiedAccessAttribute.** It could lead to more readable and more robust code. But a little bit more verbose? What do you think?

Anyway since this now works I will go on with the remaining samples.
I may have to modify the bindings again. Then do you want me to open PRs or just wait until all is done?

**Comment by alfonsogarciacaro** [comment]
Glad it's working for you 😄 

> I wonder about RequireQualifiedAccessAttribute. It could lead to more readable and more robust code. But a little bit more verbose? What do you think?

Hmm, I don't an opinion on this yet. `RequireQualifiedAccessAttribute` is not generally used in F# to allow for terser syntax. But it's true that with `KeyValueList` we are facing naming conflicts often. If you're writing the binding and see that there'll be conflicts, probably it's better to use `RequireQualifiedAccessAttribute` to prevent confusion, yes.

> I may have to modify the bindings again. Then do you want me to open PRs or just wait until all is done?

I think there're not many people yet using the Pixi bindings (hopefully thanks to your work many more will soon) so please feel free to send PRs if that's better for your workflow. Whatever is most convenient for you. I can also give you write access to the repo and npm package if you need to do a quick update :+1:

**Comment by whitetigle** [comment]
ok. Since there's no hurry (apart from my side), what I will do is go on until I'm done (19 done/21 remaining) and send links here to updated bindings so that @MangelMaxime and other people can also check them as needed.

Then eventually we will merge the changes.

As a side note I think It would be great eventually to be able to host several bindings (v3, v4, etc..) and something to automate compilation of the samples so that we can check there aren't breaking changes when we push changes to both the bindings and the samples. Would that be possible?

Then for the samples I would like to replicate the system used by pixi web site: choose the version of the lib and so load the samples using the relevant f# bindings. (check http://pixijs.github.io/examples/index.html?s=demos&f=movieclip-demo.js&title=MovieClip#/demos/render-texture-demo.js at the botton of the page)
Well that's some food for thought!

Anyway thanks for your feedback!

**Comment by whitetigle** [comment]
Just modified the bindings to allow for optional masking in Container

``` f#
member __.mask with get(): U2<Graphics, Sprite> option = failwith "JS only" and set(v: U2<Graphics, Sprite> option): unit = failwith "JS only"
```

(updated bindings here, same url: https://gist.github.com/whitetigle/b03e4689823d2950b711a81d4793df04)

I changed this in order to be able to manage such case:

``` javascript

stage.on('click', onClick);
stage.on('tap', onClick);

function onClick()
{
    if(!container.mask)
    {
        container.mask = thing;
    }
    else
    {
        container.mask = null;
    }
}
```

since **None** become **null** in JS, and since pattern matching allows for cleaner code I went up with  the following:

``` f#
let onClick() =
  match container.mask with
  | None -> container.mask <- Some(U2.Case1 thing)
  | Some(mask) -> container.mask <- None
  ()

stage?on("click", fun e -> onClick())
stage?on("tap", fun e -> onClick())
```

Is this Ok?
Thanks!

**Comment by MangelMaxime** [comment]
I think it's ok.

You can remove the `()` around `mask`.
And I think, you can also remove the final `()` because your match is already returning unit.

``` fsharp
let onClick() =
  match container.mask with
  | None -> container.mask <- Some(U2.Case1 thing)
  | Some mask -> container.mask <- None
```

However, I didn't test the code.

**Comment by alfonsogarciacaro** [comment]
@whitetigle, you and @MangelMaxime are now the Fable-Pixi experts so anything you decide will be fine :) 

About managing versions of the bindings, the truth is I haven't thought anything about this and it'd be good to have something. However, in this case it may be difficult to maintain support for several versions of Pixi. You mentioned above that it's a bit early to use v4 but as you plan to port your app anyways to F#, couldn't we start supporting Pixi v4 and solve the problems as they appear?

**Comment by MangelMaxime** [comment]
From my point of view, here are the questions that we need to answer.
1. So the first question is should we support both version side by side ? Or just the last one ?
2. How can we deliver the two versions ? 
   - Two different packages ?
   - Both binding in the same with a different name ? Fable.Import.Pixi.V3.js vs Fable.Import.Pixi.V4.js 

For my part, I think should try to focus on the v3 of pixi for the moment and structure the repo from it. Like adding the samples, build system, gh-pages for tutorials or documentation,  etc.

Then when we have something nice with v3, we could extends it to v4. But there we will have the same problem about managing the version :).

If we take @whitetigle 's samples, we should convert his build system to [FAKE](http://fsharp.github.io/FAKE/) one I think which his more common for F# projects. I could give it a try if you are ok ?

**Comment by alfonsogarciacaro** [comment]
Yeah, I forgot about that. FAKE is indeed the obvious choice to automate the compilation of the samples. Fable's `build.fsx` contains some Node and Npm helpers if you want to have a look. Of course you can ask me if you have any doubt :)

About the bindings version, we could use `package.json` for that. For example, bindings for Pixi v3 or v4 would be numbered 3.0.x and 4.0.x respectively.

**Comment by MangelMaxime** [comment]
The `package.json` could be a good solution yes. 
Does npm support updating any packages version ? I mean doesn't the each update be increasing the version of the package ?

And so the repo would be architecture like ?

```
pixi
+--V3
|    +-- Fable.Import.Pixi.fs
+--v4
|    +-- Fable.Import.Pixi.fs
```

**Comment by whitetigle** [comment]
Great news guys if we can have both versions and an automated build system! I think it will prove of unvaluable help for devs. :+1: 

For instance check this sample: http://pixijs.github.io/examples/#/demos/masking.js
Change the version of the lib (bottom of the page) and you will see that the behavior is very different between v3 and v4 (not to say a little bit scary...)

And ok to focus on v3 first like @MangelMaxime said and begin to build things on top of this version and then provide with v4 soon after and start speaking about it. 

Regarding the samples, I will continue with side-by-side comparison: straight port that looks like as much as possible as the JS version so users can jump in the boat right away.

But eventually what would be great would be to have _**F# powered**_ versions of the samples that unleash fonctional and language enhancements. So it would be really cool to review the samples once done and see how we can make them more F#ish and eventually offer the choice for users.

Side note:
I must say that porting the samples have proven easier than I thought and the generated code remains light and easily readable! Moreover Ionide is great to work with and compilation is quite fast. And so far I like the language which is concise. So for me that's some real good news. :+1:

**Comment by whitetigle** [comment]
Regarding the npm package, couldn't we have a single package with both bindings and then let the user choose in the path the most relevant one? Ok, it will download more things but I think that it would not be difficult to understand. (But maybe I fail to grasp some specifics regarding npm?)

Example:

```
#load "../../node_modules/fable-import-pixi/v3/Fable.Import.PIXI.fs"
or
#load "../../node_modules/fable-import-pixi/v4/Fable.Import.PIXI.fs"
etc...
```

like @MangelMaxime idea?

**Comment by alfonsogarciacaro** [comment]
### Version management

@MangelMaxime, to be honest, I'm actually not sure if all the versions released in npm have to be incremental but I don't think so as there are good cases for keeping two major versions alive and at least I know it's possible to make prereleases of the next major version in parallel with the current one. I'm also not against @whitetigle's proposal of putting both versions in the same package. This is just no usual in other Fable bindings, but I guess if the README includes proper instructions that should be fine :)

> Please also note that at the moment there's no way to enforce matching of the Fable bindings and the actual JS lib, so it'd be responsibility of the user to make sure they're using the binding with the proper Pixi version.

In any case, for these matters I'm always thinking of **maintenance**. If you compromise to maintain the bindings (with my help of course) whatever method works best for you will do :wink: I have given you both access to this repo and if you register in npm, I can also make you collaborators of the fable-import-pixi package so you can update it in case of need.

@whitetigle, I'm really happy you're enjoying programming in F#. It's a great language and in Fable we strive to get the balance between functional programming correctness and pragmatics (e.g. allowing dynamic programming, making bindings easier). Of course when you're finished with the examples I can review them and give you suggestion if some parts can be made more expressive or idiomatic in F# :+1:

**Comment by MangelMaxime** [comment]
My only concerns about having both version of the binding in the same package version. Is if we fix one of the binding we increment the both version "virtually speaking".
So if we want to track the changes in the binding with RELEASE notes. It's will not be easy for the user to understand why is version the package have changed but with no change in v3 for example.

I would recommand to try having two major version of the package like @alfonsogarciacaro proposition.

> About the bindings version, we could use package.json for that. For example, bindings for Pixi v3 or v4 would be numbered 3.0.x and 4.0.x respectively.

And if we see that we can't maintain 2 versions of the repo then we could go for both binding at once in the repo.

@alfonsogarciacaro Thanks for the granted access i will try to push a minimal structure of the repo today. With docs generation etc.

**Comment by whitetigle** [comment]
> I would recommand to try having two major version of the package like @alfonsogarciacaro proposition.

As long as we can maintain both that's ok for me although I think that's more work than really needed for starters. Moreover Pixi team does not seem to want some **"Python v2.x.x VS Python v3.x.x"** like issue so eventually everybody will end up using the latest version. 

But to paraphrase @alfonsogarciacaro said:

> In any case, for these matters I'm always thinking of maintenance. If you compromise to maintain the bindings (with my help of course) whatever method works best for you will do :wink:

@alfonsogarciacaro Thanks too for the access :+1:  Here is my npm nickname: **whitetigle** (https://www.npmjs.com/~whitetigle)

**Comment by whitetigle** [comment]
> Moreover Pixi team does not seem to want some "Python v2.x.x VS Python v3.x.x" like issue so eventually everybody will end up using the latest version.

Reading myself, I am not being really clear. 

What I mean is that Pixi team will always push people to go for the latest release while allowing to stick to a previous one. So do we really need to build one repository for every version? 
Let's say we have just one, so when we go v5, users will just have to upgrade the package and change the version number in the code. 
They won't have to download another package to try, change their config, build system,etc... 
And if v5 does not work for them, then they will be able to just stick to v4 without doing any **"administrative"** task. Basically it would be a simple "flag". Nothing daunting and the only command ever used then would be: `npm update`.

But I agree with the fact that the package number would only reflect **changes** to any binding, not necessarily  **improvements** of the latest pixi bindings. But with a RELEASES doc, we could make that clear?

Anyway I'll stick to your choice. So it's up to you.  :smile:

**Comment by MangelMaxime** [comment]
Here is my npm nickname: [mangelmaxime](https://www.npmjs.com/~mangelmaxime)

I think we are thinking too much.

Here is my plan:
1. Go to work with v3 for the moment because we are both working with this one at the moment
2. In the future, update the binding for supporting v4 by making a new package major version 4.x.x

And that's it. As you said pixi team will try to push the user to use the latest version and that make more sense. So we should make active support for the latest too (when ready to support it^^).

I hope I am clear with my proposition...

And like @alfonsogarciacaro said if we need to patch the v3 of the binding then the package 3.x.x should allow us to fix this one. Otherwise how could the standard libraries maintain their "old" major version.

**Comment by whitetigle** [comment]
@MangelMaxime :+1: 

Did new changes in the bindings to allow smoother use of ParticleContainerProperties:

``` f#
type PCP = ParticleContainerProperties
let options = [
    PCP.Scale true
    PCP.Position true
    PCP.Rotation true
    PCP.Uvs true
    PCP.Alpha true
  ]
let sprites = new ParticleContainer(10000.,options)

```

(For now, updated bindings still there: https://gist.github.com/whitetigle/b03e4689823d2950b711a81d4793df04)

**Comment by whitetigle** [comment]
@alfonsogarciacaro 

>  in Fable we strive to get the balance between functional programming correctness and pragmatics (e.g. allowing dynamic programming, making bindings easier). 

Yeah! Really enjoyed this syntax this morning: 

``` f#
dude?turningSpeed <- Math.random() - 0.8
```

Dynamic programming indeed :smile: 

>  Of course when you're finished with the examples I can review them and give you suggestion if some parts can be made more expressive or idiomatic in F# 

That would be great indeed! Thanks!

**Comment by alfonsogarciacaro** [comment]
Is it the API for Pixi v4 very different from v3? [Here](http://www.goodboydigital.com/pixi-js-v4/) says the API is 99% maintained so maybe it's not worth to keep two different versions for the bindings. Maybe we can have just one and mark types belonging only to v4 with a comment.

Also, to have two different bindings we should have two different typescript d.ts declarations for Pixi v3 and v4. [Here](https://github.com/pixijs/pixi-typescript/blob/master/pixi.js.d.ts) I can only find the d.ts for v3.

**Comment by MangelMaxime** [comment]
For the d.ts there is a branch [v4.x](https://github.com/pixijs/pixi-typescript/tree/v4.x)

Yes we could mark them by a comment but I think sometimes the type name will probably be the same but with different methods etc.

But yes, in the long our idea is to support only the v4. Because if someone want to use pixi with fable then he will probably start a new project. So this not a bad idea to "force" him to go for v4.

**Comment by whitetigle** [comment]
Hi!
I was working on porting this sample: http://pixijs.github.io/examples/#/demos/texture-rotate.js and ran into some issues with the current bindings so here is an update:
- Added a missing GroupD8 structure
  source js: https://github.com/pixijs/pixi.js/blob/master/bin/pixi.js#L13345
  source ts v3: https://github.com/pixijs/pixi-typescript/blob/master/pixi.js.d.ts#L347
  source ts v4: https://github.com/pixijs/pixi-typescript/blob/v4.x/pixi.js.d.ts#L564
- corrected type for rotate parameter in Texture constructor
  source ts v3: https://github.com/pixijs/pixi-typescript/blob/master/pixi.js.d.ts#L1132

(For now, updated bindings still there: https://gist.github.com/whitetigle/b03e4689823d2950b711a81d4793df04)

**Comment by whitetigle** [comment]
Ok I'm done with the Demos section. So I think we covered a lot of the basic API.

Now there's a remaining 10 samples batch which dive further into more specialized stuff (Spine & DirectDrawSurface for instance).

Through the samples, I noted different results between v3 and v4. Latest one being: http://pixijs.github.io/examples/#/demos/texture-rotate.js

So I think that will be important to allow users to test the samples with v3 and v4. I also think we should link to the official samples so that if any doubt arise, people don't say it does not work as expected because of our bindings. If you have any other idea to deal with that kind of stuff, please share it :smile: 

I am also thinking about splitting the window into 2, to allow a F# VS JS direct comparison (will need to display line numbers). What do you think?

**Comment by alfonsogarciacaro** [comment]
Fantastic work! :clap: Can you please push the updated bindings to fable-graphics repo? About publishing the samples, I think we should start simple and go step by step depending on user response:
1. If you don't mind, we could move your pixi samples to this repo (with attribution, of course) so users can find them more easily.
2. Make a couple of samples live in Fable's website (indicating which Pixi version we're using)
3. Then, create a custom websites for fable-graphics will all the samples as with [fable-virtualdom](https://fable-compiler.github.io/fable-virtualdom/).
4. There's already work going on to create a [Fable REPL](https://github.com/fable-compiler/Fable/pull/368). When this is finished, we can adapt it to create a REPL for Pixi samples (so users can edit them online).
5. When this is done we can think about giving the option to run the samples against different Pixi versions :wink:

**Comment by MangelMaxime** [comment]
For information, I am actually trying to port the docs from fable-virtualdom into this repo.

When done we should be able to import the samples from @whitetigle repo if he is ok.

**Comment by whitetigle** [comment]
@alfonsogarciacaro made a diff between current bindings and mine and tried to push new but was denied access. **remote: Permission to fable-compiler/fable-graphics.git denied to whitetigle.**

@MangelMaxime Excellent! We'll do that!

**Comment by whitetigle** [comment]
@alfonsogarciacaro 

> Make a couple of samples live in Fable's website (indicating which Pixi version we're using)

As a single page?

**Comment by MangelMaxime** [comment]
For the sample Yes I think we should go for a single page with 1-3 samples. Just as a show case on the Fable website.
For example, the virtual-dom page contain 3 samples.

**Comment by alfonsogarciacaro** [comment]
> made a diff between current bindings and mine and tried to push new but was denied access. remote: Permission to fable-compiler/fable-graphics.git denied to whitetigle.

Github says the invitation I sent you to join `fable-graphics` team is pending. Can you please check you mail and accept it? I think if you to [fable-compiler org page](https://github.com/fable-compiler) you should be able to see and accept the invitation too.

> For the sample Yes I think we should go for a single page with 1-3 samples. Just as a show case on the Fable website. For example, the virtual-dom page contain 3 samples.

I think we're talking of different things 😄  For [Fable's websites samples](https://fable-compiler.github.io/samples.html) we could add 1 or 2 Pixi samples, I would add one, tweet it and then add the second one (in a different page) a few days later. This helps keep people interest alive :wink:

In the fable-graphics site @MangelMaxime (in the same fashion as [fable-virtualdom site](https://fable-compiler.github.io/fable-virtualdom/)) we can add as many Pixi samples as we want :smile:

**Comment by whitetigle** [comment]
> Github says the invitation I sent you to join fable-graphics team is pending. Can you please check you mail and accept it? I think if you to fable-compiler org page you should be able to see and accept the invitation too.

Oops did not see that mail. Ok I'm in and I just pushed the updated bindings.

> I think we're talking of different things :smile: For Fable's websites samples we could add 1 or 2 Pixi samples

So if I understand well @MangelMaxime is preparing the structure of the project that will eventually automate the builds, the doc, etc... And allow to display all the samples (custom website).

On my side, I have to finish to port the remaining samples and prepare a few interesting ones with complete instructions to publish on Fable's site so @alfonsogarciacaro can work on communication and bring to us hordes of interested developers!

A third step will be to make our bindings available for some online fun (REPL) and then, eventually, go into the details of managing multiple PIXI version.

Is that correct? Please forgive me if there's something I did not understand. 
Thanks!

**Comment by whitetigle** [comment]
As a side note. My last push has been made with the name halfabench. Don't know why this name's still linked with my account. I will check. Anyway just so it's clear halfabench is me :smile: 

[EDIT] forgot to mention PIXI in the commit message. Will do next times. sorry.

**Comment by whitetigle** [comment]
Hi!
Just ran into several things I am confused with. So I need your help to understand.

**1 - new or not new?**
What's the difference between 

``` f#
let myContainer = Container()
and 
let myContainer = new Container()
```

Both seem to work the same way.

**2 - mutability**

``` f#
let container = Container()
container.filters <- Some(ResizeArray[|myFilter :> AbstractFilter|])
```

Does it mean every field is mutable?

Thanks for your help here! And sorry if it's already in the docs.

**Comment by whitetigle** [comment]
Just went through the whole filters section (http://pixijs.github.io/examples/#/filters/displacement-map.js) and updated the bindings to allow for null filters member in DisplayObject.

**Comment by alfonsogarciacaro** [comment]
> 1 - new or not new?

In F# it's actually the same, they usually recommend to use `new` with objects implementing `IDisposable` as a visual tag.

> 2 - mutability

`ts2fable` makes all properties in the bindings as mutable as there's no way to utomatically check in TypeScript declarations which properties are supposed to be mutable or not. Properties in types/modules defined in F# code will be immutable unless marked explicitly with `mutable` (or by adding a `set` method).

**Comment by whitetigle** [comment]
@alfonsogarciacaro thanks!

**Comment by whitetigle** [comment]
All **pure pixi** samples done!

Notes:
- One sample in the picture section does not work (as written in its comments) the other one needs a plugin
- **DisplayList**: samples need appropriate plugins
- **Spine**: samples need appropriate plugins
- **DirectDrawSurfaces**: samples need appropriate plugins

**Comment by whitetigle** [comment]
Pixi being in the process to work, I will need 
- Howler.js for sound
- a tweening library

So, eventually I will have to port Howler.js too :)
Regarding the tweening lib, do we have one we can already use or should I go for a JS one?

**Comment by alfonsogarciacaro** [comment]
By tweening you mean purely math operations? It that's so, maybe there's something for in F#/.NET we could translate, but I'm not sure. If it requires something more platform related (like image manipulation), it'll be difficult to translate and it's probably better to find a JS lib.

**Comment by whitetigle** [comment]
OK. I'll see. Thanks!

**Comment by alfonsogarciacaro** [comment]
@whitetigle I reviewed the samples, great job! :) I've made some small changes, mainly about formatting (shortening lines so they fit better in the web, etc), in the hope they will make the code a bit clearer for other users. You can check them and ask me if you have any question :) 621e856072a0e8eaad63053ab7f11ce75ddce90a

**Comment by whitetigle** [comment]
@alfonsogarciacaro thanks! Cleaner is better! :+1: 
And I can now remove "check & clean mutable" from my todo list :wink: 

Question : what purpose lies behind the redesign of the **animate** function? 

[EDIT]

You wrote a few days ago;

> This was actually a bug in Fable, it should be possible to type test erased union types (e.g. match x with U2<string,int>.Case1 s -> ... will translate to typeof x === "string" ? ...). It will be fixed in the next release

so does this sample mean we're good to go? 

```
let isWebGL, renderer =
  match Globals.autoDetectRenderer(800., 600.) with
  | U2.Case1 webgl -> true, webgl :> SystemRenderer
  | U2.Case2 canvas -> false, canvas :> SystemRenderer

```

Also what does the **false,xxx** syntax mean?

`U2.Case2 canvas -> false, canvas :> SystemRenderer`

Thanks!

**Comment by alfonsogarciacaro** [comment]
@whitetigle The refactoring of `animate` function is a common pattern to keep the state (in this case, the `count` variable) of the function private. `count` now is trapped in a closure by the inner `animate` but it's not accessible from the outside so you can be sure nobody else is modifying the value. I hope that makes sense 😄 

And yes, the bug for making pattern matching with erased unions is fixed :+1: Thanks a lot for discovering it! In the sample above, returning a tuple and then destructuring in the assignments is a common pattern in F# to return two values in one go (`webGL` is the boolean value, and `renderer` the result of `autoDetectRenderer` upcasted to `SystemRenderer`).

**Comment by whitetigle** [comment]
@alfonsogarciacaro Thanks for these explanations! Everything's quite clear!

Now I have been playing a little bit to port one of my project. 
So I started with a simple structure with the following rules:
1. **the state of the game can change.** For instance after the init animation (Init state), I want to enter the Play state. Then if I die, the GameOver state. each state is responsible for it's own logic and rendering.
2. while I can derive a state from the previous one **I need a way to make it evolve as needed**. For instance after GameOver, I want to be able to choose a state relative to what happened in the game.

Here a sample which is directly applied from one of my very small project. 

``` f#
type GameStates =
    | Init
    | LevelUp
    | Play
    | GameOver of GameStates

let rec animate =
   let mutable state = Init
   let rec animate (dt:float) =
     count <- count + 1.

     match state with
        | Init ->
          printfn "Init"
          if count > 100. then
            state <- Play

        | Play ->
          printfn "Play"
          match Math.random() * 10. with
            | x when x < 2. ->
              state <- GameOver Init
            | _ ->
              match count with
                | x when x > 100. ->
                  state <- LevelUp
                | _ -> ()

        | LevelUp ->
          printfn "LevelUp"
          count <- 0.
          state <- Play

        | GameOver nextState ->
          printfn "gameover"
          count <- 0.
          state <- nextState

     renderer.render(stage)
     window.requestAnimationFrame(FrameRequestCallback animate) |> ignore
   animate

  // start animating
animate 0.

```

My main problem is since `window.requestAnimationFrame` needs a `float->unit` how can I actually pass a State and then pass an updated copy so we remain into the immutable realm. 
Is there any functional design pattern that would match this case? Is this possible?

Thanks a lot for your help!

**Comment by MangelMaxime** [comment]
@whitetigle I am not sure but I think the architecture used in the project https://github.com/fable-compiler/fable-virtualdom can be a solution.
I have personally tried it at the beginning when working GUI lib over Pixi and it's was promising.

The idea behind Fable Architecture is that we react over Message that with use to update the Model and when the Model has been updated with can Draw the view from the Model.

When playing with the architecture I needed something like 50-60lines to support manage the application. I don't know if it's what we should go with for games. Perhaps wait for the answer of Alfonso

**Comment by whitetigle** [comment]
@MangelMaxime thanks for the hint! I will definitely take a look.

**Comment by alfonsogarciacaro** [comment]
@whitetigle It'd be nice to adapt the architecture developed by @mastoj and @MangelMaxime to a game engine (aka Fable architecture :wink:). I've to admit I didn't have much time to play with it myself so I'm not sure if it's a good fit or not.

Probably @MangelMaxime can correct me, but my concern is the Fable architecture (as well as other architectures designed for business applications like Elm or Redux) is event-driven (the updates are triggered when an event like a button click happens in the GUI) but in a game loop updates happen for every frame.

At a very basic level what you need is an **asynchronous recursive** function so you can feed the function with the updated state instead of storing a copy somewhere and keep everything immutable and nice 😸 This can be done with F# `async` computation expression. For example:

``` fsharp
// Convert requestAnimationFrame to Async
let awaitAnimationFrame() =
  Async.FromContinuations <| fun (cont,_,_) ->
    window.requestAnimationFrame(FrameRequestCallback cont) |> ignore

let rec animate state = async {
  let state = update state // Update the state
  let! _ = awaitAnimationFrame() // Wait for next frame
  return! animate state // Recursive call
}

animate initState |> Async.StartImmediate
```

Maybe this meets your needs. I've modified the blur-filter sample in the [async-recursion](https://github.com/fable-compiler/fable-graphics/commit/91d9b4dfd0c61ea095d1a72e064eb958c8c768af) branch to check the code works... and it does :+1:

**Comment by MangelMaxime** [comment]
You are true about that Fable architecture is event-driven. But it's can also react to subscriber in order to push an update. [Here](https://fable-compiler.github.io/fable-virtualdom/samples/clock/index.html) is an example of a clock.

I think this should be sufficient for certain type of games for example, this [tetris game](https://github.com/w0rm/elm-flatris) is made using Elm and the Elm architecture. (Fable architecture is actually a port from the elm one).

Another solution is to follow the implementation shown by Elm games samples for example this mario game ported to Fable: https://fable-compiler.github.io/samples/mario/index.html

But here the abstraction is really thin.

There are others solutions existing in Elm, Purescript etc. but I find the two quoted above the most versatile and powerful. Because both enforce you to structure the program over a model. 

Initial Model -> Update Model -> Render Model -> Update Model -> Render Model -> ...

And I find this structure really nice and you can build a lot of great tools over it to help you. Like time travel, watcher, debug (what again a time travel :) )

**Comment by alfonsogarciacaro** [comment]
Right! I missed the part about producers in Fable architecture, sorry 😅

**Comment by MangelMaxime** [comment]
No problem :) I work with it all the day and night that help a little :)

**Comment by whitetigle** [comment]
First of all, thank you both for your help. I begin to get a clearer and broader vision of F# and Fable. And I like it a lot :smile:  

So for starters I tried a first approach with the async idea from @alfonsogarciacaro.

``` f#
type GameState =
  {
    Count: float;
    Level: int;
  }

type State =
    | Init of GameState
    | LevelUp of GameState
    | Play of GameState
    | GameOver of State * GameState

let awaitAnimationFrame() =
  Async.FromContinuations <| fun (cont,_,_) ->
    window.requestAnimationFrame(FrameRequestCallback cont) |> ignore

let rec animate state = async {
  let state =
    match state with
    | Init state->
      printfn "Init"
      let updatedState = {state with Count = 0.; Level = 0}
      Play updatedState

    | LevelUp state ->
      printfn "LevelUp"
      let updatedState = {state with Level = state.Level + 1}
      printfn "Level is now %i" updatedState.Level
      Play updatedState

    | Play state ->
      printfn "Play"
      let updatedState = {state with Count = state.Count + 0.1}
      match updatedState.Count with
      | x when x > 100. ->
        let stateAfterGameOver = Init updatedState
        GameOver (stateAfterGameOver, updatedState)
      | value ->
        let nextLevel = (Math.floor(value) |> int) % 2 = 0
        if nextLevel then
          LevelUp updatedState
        else
          Play updatedState

    | GameOver (nextState, gameState) ->
      printfn "GameOver"
      printfn "Level is %i" gameState.Level
      nextState

  renderer.render(stage)
  let! _ = awaitAnimationFrame()
  return! animate state
}

// start animating
let startingState = {Count=0.; Level=0}
let StartState = Init startingState
animate StartState |> Async.StartImmediate
```

So it's a basic loop that updates a GameState record.
The game begins (**Init**) then enters it's core logic **Play**. Every few frames, we increase the level (**LevelUp**). After a while the game stops (**GameOver**) and comes back to **Init**.

There's the render call for the current frame that does nothing in this sample but would update sprites, effects, etc..
There aren't any  event. And I haven't taken time yet to check how they would fit.

So it is working, it looks like what I was looking for and I think it's a good basis for discussion. 
So now @MangelMaxime do you think this could somehow be adapted to Fable architecture?

**Comment by MangelMaxime** [comment]
@whitetigle Yes this could be adapted to Fable architecture. I will try to implement a simple version with your sample just to have a comparison.
I keep you inform.

**Comment by alfonsogarciacaro** [comment]
I must also say that I overengineered my solution above a bit. The same effect can also be achieved without `async` in this case 😅 

``` fsharp
let rec animate state =
  let state = update state
  renderer.render(stage)
  window.requestAnimationFrame(FrameRequestCallback(fun _ -> animate count)) |> ignore

animate initState
```

**Comment by whitetigle** [comment]
`FrameRequestCallback(fun _ -> animate count)` Well done! I updated my code.

**Comment by whitetigle** [comment]
@MangelMaxime I was wondering this morning about having the ui components in the dom to leverage the virtual dom approach and let the core game rendering with requestAnimationFrame as usual.
Now the questions is how would they communicate. Through a shared Model?

**Comment by MangelMaxime** [comment]
@whitetigle VirtualDom support what we call Producers and Subscribers.

So if you want to communicate with the external world you can use them. For example, this is how we are managing the SPA approach for the router part (the browser detect the `hashchange` and push the information in the app).
- A producer is something that push message into your application.
- A subscriber is a function which will be notify by the application on every message handle by the application.
  And as everything that happen in the application is handle by a message you can listen what ever your want.

Here is a sample of a Producer: https://github.com/fable-compiler/fable-virtualdom/blob/master/samples/clock/clock.fsx#L68

Here is a sample of a Subscriber: https://github.com/fable-compiler/fable-virtualdom/blob/master/samples/subscriberexample/subscriberexample.fsx#L54

Also, in a future version of VirtualDom we will have support for a plugin structure which will allow us to capture not only the messages but also the Model. 
[Here](https://github.com/fable-compiler/fable-virtualdom/blob/feature/plugin/samples/plugin/plugin.fsx) is the current sample of the plugin usage. :warning: **Still under development**

At the current point I think Producer/Subscriber can be enough because I don't think we want to share the whole model between the Pixi world and VirtualDom world. The Model will have more information than just the useful one like HP amount, clicked, etc. For example, in VirtualDom you need to keep track of the value in the input.

And as I am working on the Fable Architecture version for Pixi binding we could probably have a Pixi App and a VirtualDom App. I hope to publish a first version today.

_Off scope of your question:_
_Finally, I understand that having the GUI through DOM context could be a big improvement but I don't like CSS and HTML. So I am working on a Library which will allow to have a full web app working inside a Canvas. So later we could also offer this solution to users too._

**Comment by MangelMaxime** [comment]
@whitetigle I have pushed on a new [branch](https://github.com/fable-compiler/fable-graphics/tree/feature/fable-architecture/samples/pixi/fable-architecture) a sample showing how we could implement the Fable Architecture over Pixi.

I like this kind architecture which enforce the separation between the View / Model / Update. But here the view function is quite hacky because we don't have the support of VirtualDom / React to help us handle the diff, patch, rendering etc. 
However this can give us a good start to see if this kind of architecture is suitable for us or not.

I think you have a better experience of Pixi than me and probably you will another way of handling the view. I don't really know how we use Pixi to do that :) (learning it since 1 week^^)

I put a lot of comment to help understand what's going on but if you have a question don't hesitate to ping me.

@alfonsogarciacaro Ping you, if you want to take a look.

From my point of view, this architecture could be viable if we find a better way to handle the View part. If you have any idea I am open :)
Otherwise, we could try to play with your last proposition to enforce the architecture.

Perhaps we should try to make a really simple game like pong with both and try to see which one fit the best our needs.

**Comment by whitetigle** [comment]
Well just a few thoughts about pixi before I dwelve into the details of your sample (very impressive by the way :+1:  ) :

We do need **requestAnimationFrame** since that's the only method that guarantee that the next draw will happen as soon as possible. It's kind of _Browser enforced_.

**SetInterval** tries as fast as it can but can be delayed by system calls or whatever "noise" so there's absolutely no guarantee we'll get a stable frame rate.

Then we would have different needs:
- **slow rendering objects**: for instance **UI** that do not need to update as fast as possible. In most cases, UI is controlled through tweening which is independent from the render loop speed since it's time based. The render just needs to refresh the screen with the updated values. That would be a good candidate for our tests.
- **fast rendering objects**: particles, and all the objects that need fast updates on the screen. Pixi takes care of them and is already optimized so there's no need to update them outside of the render loop.

Then we do have events that need to interact with both kind of objects. 
For UI no problem, we usually bind events to UI components and voilà. But sometimes UI controls fast rendering objects: increase speed, etc... So we need to be able to communicate.

For fast rendering we have two scenarios:
1. bind events to the objets. For instance click on a foe to kill him.
2. bind events to the stage to detect and interpret player behavior

There too Pixi takes care of the events for us and we want to communicate as fast as possible to avoid any lag.

Now all games are different. 
- A match3 would need slow rendering apart for particle effects.
- Pong would mostly need fast rendering for display and events
- mario would need a very stable framerate
- a Schmup would need a very stable and fast framerate too 
- a click & play would allow for very interesting tries since it's usually playing anims at a given framerate but not faster (after all we don't want [Guybrush Threepwood ](https://en.wikipedia.org/wiki/Guybrush_Threepwood) to run through a pirate! )

So most of the games want 40-60 fps and they want it smooth (hence the need for requestAnimationFrame). But there's an outstanding number of games that do not need more. 
For the others (FPS for instance) I think we're out of scope. Unreal Engine is their friend. 

So there are several choices we could make that would just be pragmatic and allow us to ask for fast rendering just when it's needed, not for all components, not all the time if it allows us to have a good and robust architecture. 

So I will dive into your code and try to understand as much as I can. 
Thanks for your work :+1:

**Comment by MangelMaxime** [comment]
Thanks for all this explanations and this is what I what thinking when writing the sample. 

For me we should dig in the direction of your proposition and the mario architecture. And perhaps simply offer the user helper library like Keyboard event management, tween, etc. 

But not "forcing" them to Fable Architecture or a similar one has it's probably to complex for a game. We could probably simply say here is the recommended way to go with Fable + Pixi and let them experiment they own architecture if they want.

I have always hated play with Engine because they force you to do things their way and usually I don't like it :).

[Edit]: I will wait your impression on the work done. And perhaps we could simply take inspiration from part of the work.

**Comment by alfonsogarciacaro** [comment]
@MangelMaxime Great work! I've limited access to my computer until Friday so I cannot have a deep look yet but I'll check later to see if I can so some comment (though I think you've now much more expertise on Fable architecture than me :wink:). In any case, I agree with @whitetigle that `requestAnimationFrame` should be used. I think Elm uses it now for DOM rendering too, and it has several niceties: the browser automatically adjust the frame rate to the device power, pauses the animation if the tab isn't visible, etc.

**Comment by MangelMaxime** [comment]
Yes I agree too about the requestAnimationFrame just did think about it on the moment. I will change the producer to use this function.

Probably true about Fable architecture but you still have a better understanding of  the Functional paradigm :) So you could probably think of a better way to handle the view function which is not has good has it should be ^^. (But I didn't find a good structure).

**Comment by whitetigle** [comment]
Ok. I would like to make a test with porting bunnymark http://www.goodboydigital.com/pixijs/bunnymark_v3/ to your architectrue and see what happens.

What do you think? Could you just update the code so we use `requestAnimationFrame`? I am not sure how to make that change in your code. Thanks!

**Comment by whitetigle** [comment]
Hi!
I just pushed a little pixi demo there:  [pixi |> fable](http://fable.francoisnicaise.fr/sample.html?s=lab/dark-fable&p=out/lab)

If it's ok for you, Since it's quite a complete sample, I plan to add it to the samples when done with the inline documentation.

Featuring:
- Graphics API
- Sprites
- textures
- masks
- events
- filters
- bitmap texts
- Containers
- and some easy to use maths

So a kind of Pixi soup :wink: 
I need to rework a bit the layout to better fit 16:9 displays.  But it works fullscreen and is even nicer in windowed mode. 

To change the speed of the animation, just move your mouse close or far from the center of the screen. farther = faster. closer = slower. Mouse up and down to fade the animation et voilà!

Last but not least, I will push some new changes to the bindings this week since I have found new little optimizations.

**Comment by alfonsogarciacaro** [comment]
@whitetigle That looks über cool! :clap: https://twitter.com/FableCompiler/status/775310166988324865

**Comment by MangelMaxime** [comment]
@whitetigle It's a really cool sample. I love it :) Nice work :clap:

**Comment by whitetigle** [comment]
Thanks! :smile: I'm glad you enjoy it! 

Oh, by the way, Uuntil I'm ready to push the sample, the source is [there](https://bitbucket.org/fnicaise/fs_press/src/7bba6c90029538843d1cf930e9bd4b74ed0c8354/src/lab/dark-fable.fsx?at=default&fileviewer=file-view-default)

One more thing, @alfonsogarciacaro  since you use twitter, and I can't get access anymore to my old account @whitetigle, I have created an new account: @thewhitetigle. 
So if you wish to link to my account, please use the new nickname. Thanks!
